dkg/frost/frost.go

```diff
@@ -0,0 +1,277 @@
+}
+
+type DKGRound int
+
+const (
+	Preparation DKGRound = iota + 1
+	Round1
+	Round2
+	// Blame
+)
+
+func New(
+	requestID dkg.RequestID,
+	network dkg.Network,
+	i, t uint32,
+	operators []uint32,
+) dkg.KeyGenProtocol {
+
+	otherOperators := make([]uint32, 0)
+	for _, operatorID := range operators {
+		if operatorID == i {
+			continue
+		}
+		otherOperators = append(otherOperators, operatorID)
+	}
+
+	ctx := "string to prevent replay attacks"
+	party, _ := frost.NewDkgParticipant(i, uint32(len(operators)), ctx, thisCurve, otherOperators...)
+
+	return &FROST{
+		identifier: requestID,
+		network:    network,
+		operatorID: i,
+		threshold:  t,
+		operators:  operators,
+		party:      party,
+	}
+}
+
+func (fr *FROST) Start(init *dkg.Init) error {
+	return nil
+}
+
+func (fr *FROST) ProcessMsg(msg *dkg.SignedMessage) (bool, *dkg.KeyGenOutput, error) {
+	protocolMessage := &ProtocolMsg{}
+	if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+		return false, nil, errors.Wrap(err, "could not decode protocol msg")
+	}
+
+	if fr.msgs[protocolMessage.Round] == nil {
+		fr.msgs[protocolMessage.Round] = make(map[uint32]*ProtocolMsg)
+	}
+
+	fr.msgs[protocolMessage.Round][uint32(msg.Signer)] = protocolMessage
+
+	switch protocolMessage.Round {
+	case Preparation:
+		if fr.canStartRound1() {
+			// do round 1
+		}
+	case Round1:
+		if fr.canStartRound2() {
+			// do round 2
+		}
+	case Round2:
+		if fr.hasFinishedRound2() {
+			// return keygen output
+		}
+	}
+	return false, nil, nil
+}
+
+func (fr *FROST) canStartRound1() bool {
+	canStart := true
+	// check if we have public keys of all participants
+	return canStart
+}
+
+func (fr *FROST) canStartRound2() bool {
+	canStart := true
+	for _, operatorID := range fr.operators {
+		_, ok := fr.msgs[Round1][operatorID]
+		if !ok {
+			canStart = false
+		}
+	}
+	return canStart
+}
+
+func (fr *FROST) hasFinishedRound2() bool {
+	hasFinished := true
+	for _, operatorID := range fr.operators {
+		_, ok := fr.msgs[Round2][operatorID]
+		if !ok {
+			hasFinished = false
+		}
+	}
+	return hasFinished
+}
+
+func (fr *FROST) processRound1() error {
+	bCastMessage, p2pMessages, err := fr.party.Round1(nil)
+	if err != nil {
+		return err
+	}
+
+	commitments := make([][]byte, 0)
+	for _, commitment := range bCastMessage.Verifiers.Commitments {
+		commitments = append(commitments, commitment.ToAffineCompressed())
+	}
+
+	shares := make(map[uint32][]byte)
+	for _, operatorID := range fr.operators {
+		share := bls.SecretKey{}
+
+		shamirShare := p2pMessages[operatorID]
+		if err := share.Deserialize(shamirShare.Value); err != nil {
+			return err
+		}
+		fr.operatorShares[operatorID] = &share
+
+		if fr.operatorID != operatorID {
+			shares[operatorID] = share.Serialize()
+		}
+	}
+
+	r1Message := &Round1Message{
+		Commitment: commitments,
+		ProofS:     bCastMessage.Wi.Bytes(),
+		ProofR:     bCastMessage.Ci.Bytes(),
+		Shares:     shares,
+	}
+
+	r1MessageBytes, err := json.Marshal(r1Message)
```

Comment:
We can make the message implement `types.Encoder` interface and use `r1Message.Encode()` here.

Reply:
sure, I'll add this